File: cases/taint/scala/computed-property-negative/ComputedPropertyNegative.scala
package dataflowbench

object ComputedPropertyNegative {
  class Holder {
    var alpha: String = "clean"
    var beta: String = "clean"
  }

  def dfb_source(): String = { // DFB-SOURCE: computed-property-input
    "tainted"
  }

  def dfb_sink(value: String): Unit = {} // DFB-SINK: computed-property-sink

  def run(): Unit = {
    val holder = new Holder()
    val writeKey = "alpha"
    val readKey = "beta"
    val writeField = classOf[Holder].getDeclaredField(writeKey)
    writeField.setAccessible(true)
    writeField.set(holder, dfb_source()) // DFB-WITNESS: computed-property-store
    val readField = classOf[Holder].getDeclaredField(readKey)
    readField.setAccessible(true)
    readField.set(holder, "clean")
    dfb_sink(readField.get(holder).asInstanceOf[String])
  }
}

File: cases/taint/java/computed-property-negative/ComputedPropertyNegative.java
package dataflowbench.taint;

import java.lang.reflect.Field;

final class ComputedPropertyNegative {
    static final class Holder {
        int tainted;
        int clean;
    }

    static int dfb_source() { // DFB-SOURCE: computed-property-input
        return 1;
    }

    static void dfb_sink(int value) { } // DFB-SINK: computed-property-sink

    static void run() throws Exception {
        Holder holder = new Holder();
        String writeKey = "tainted";
        String readKey = "clean";
        Field written = Holder.class.getDeclaredField(writeKey); // DFB-WITNESS: computed-property-member
        written.setInt(holder, dfb_source()); // DFB-WITNESS: computed-property-store
        Field read = Holder.class.getDeclaredField(readKey);
        dfb_sink(read.getInt(holder));
    }
}
